### PR TITLE
LLS-2186

### PR DESCRIPTION
**Pull request by frits-exmachina**
- New project setup, with Livery Web SDK as base.
- Added LiveryBridge, an extendable class that can communicate with postMessages
- Added LiveryInteractive, an easy to use HTML Element that uses LiveryBridge
- Added a demo page that uses LiveryInteractive with the ability to test messages.

### CODE REVIEW

**Comment by frits-exmachina** [comment]
Pushed the latest changes we discussed.

File: src/LiveryBridge.ts [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      object.type === 'resolve' &&
+      !!object.id &&
+      !!(object as ResolveMessage<ValueType>).value
+    );
+  }
+
+  public sendCommand<ValueType>(
+    name: string,
+    id?: string,
+    arg?: string,
```
Comment (difosfor) [inline src/LiveryBridge.ts:142]: Should be: `arg: ArgType` where ArgType is a second generic type which can be `undefined` if irrelevant for this command.

File: src/LiveryBridge.ts [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    return (
+      object.type === 'resolve' &&
+      !!object.id &&
+      !!(object as ResolveMessage<ValueType>).value
+    );
+  }
+
+  public sendCommand<ValueType>(
+    name: string,
+    id?: string,
```
Comment (difosfor) [inline src/LiveryBridge.ts:141]: Using default syntax here seems like it would be easier, e.g: `id = uuid(),`.

File: src/LiveryBridge.ts [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    arg?: string,
+  ): Promise<ValueType> {
+    return this.handshakePromise
+      .then(() => {
+        const commandId = id || uuid();
```
Comment (difosfor) [inline src/LiveryBridge.ts:146]: As described above using default argument syntax is probably the best option here. Alternatively I'd suggest slightly improving this by using `??` instead of `||`.

File: src/LiveryBridge.ts [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        this.sendMessage({
+          isLivery: true,
+          type: 'command',
+          name,
+          id: commandId,
+          arg,
+        });
+
+        return new Promise<ValueType>((resolve, reject) => {
+          this.deferredMap.set(commandId, {
```
Comment (difosfor) [inline src/LiveryBridge.ts:156]: Doesn't `{ resolve, reject }` work here as well?

File: src/LiveryBridge.ts [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+          resolve();
+        },
+        reject: (error: Error) => {
+          reject(error);
+        },
+      });
+    });
+  }
+
+  protected sendMessage<T extends LiveryMessage>(message: T) {
```
Comment (difosfor) [inline src/LiveryBridge.ts:270]: I'd like to see this as: `sendMessage(type: string, properties: Record<string, unknown>)` or something like that and then just have it construct the message object: `{ isLivery: true, type, ...properties }`. Likewise I suppose we could just specify id as an optional argument and have this create the id for you if not specified and return that id to be used from where this is called? Reduce the isLivery repetition etc.

File: src/LiveryBridge.ts [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+  protected targetWindowUrl: string;
+
+  protected version: string;
+
+  constructor(
+    targetWindow: Window,
+    targetWindowUrl: string,
+    version: string,
+    messageListener?: (message: string) => void,
```
Comment (difosfor) [inline src/LiveryBridge.ts:68]: Don't pass a stringified message, just pass the message object, e.g: `message: LiveryMessage`.

File: src/LiveryBridge.ts [hunk 7]
```diff
@@ -0,0 +1,296 @@
+/* eslint-disable @typescript-eslint/no-explicit-any --- MessageEvent returns an any object. The maps we use also use any, as multiple data types are allowed. */
+
+import { uuid } from './livery-interactive/util/uuid';
+
+export interface LiveryMessage {
+  id?: string;
```
Comment (difosfor) [inline src/LiveryBridge.ts:6]: Why is this optional again?

File: src/LiveryBridge.ts [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    return (
+      object.type === 'handshake' &&
+      !!(object as HandshakeMessage).version &&
+      !!object.id
+    );
+  }
+
+  static isLiveryMessage(object: any): object is LiveryMessage {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- Incoming object is any.
+    return !!object && object.isLivery === true && !!object.type;
```
Comment (difosfor) [inline src/LiveryBridge.ts:118]: Strictly speaking you should also check that `type` is a string and even one of the types that we support. Otherwise you should throw an error specifying that the type is illegal. Likewise with all of the other checks below.

File: src/livery-interactive/LiveryInteractive.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+@customElement('livery-interactive')
+export class LiveryInteractive extends LitElement {
+  bridge!: LiveryBridge;
+
+  /**
+   * Returns a promise with the SDK's current latency in seconds.
+   */
+  public getLatency(): Promise<number> {
```
Comment (difosfor) [inline src/livery-interactive/LiveryInteractive.ts:18]: Can't the return type be inferred from below?

File: src/livery-interactive/LiveryInteractive.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Comment (difosfor) [inline src/livery-interactive/LiveryInteractive.ts:19]: Why is this necessary? As a rule, always specify why you're disabling lint rule as a comment, e.g: `// eslint-disable-next-line @typescript-eslint/no-unsafe-return -- because ..`

File: src/livery-interactive/LiveryInteractive.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+  /**
+   * Initializes the livery-interactive element.
+   * @param messageListener - Optional parameter used for keeping track of messages for debugging purposes.
+   */
+  public init(messageListener?: (message: string) => void) {
+    this.bridge = new LiveryBridge(
+      window.parent,
+      '*',
+      '0.0.1',
```
Comment (difosfor) [inline src/livery-interactive/LiveryInteractive.ts:31]: This should be `'__VERSION__'` I think?

File: src/livery-interactive/LiveryInteractive.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    );
+  }
+
+  /**
+   * Returns a promise with the current screen orientation.
+   * @param listener - Called when the screen orientation changes, with the new orientation as parameter.
+   */
+  public subscribeOrientation(
+    listener: (value: string) => void,
+  ): Promise<string> {
```
Comment (difosfor) [inline src/livery-interactive/LiveryInteractive.ts:42]: Return type can just be inferred I'd say?

**Review by difosfor — CHANGES_REQUESTED** [review]
I'd like to simplify the demo page to just show getLatency and subscribeOrientation buttons and then load a mock SDK bridge that returns random latency and orientation values. That mock can also serve as a blueprint for the SDK.

File: src/LiveryBridge.ts (hunk 7 above)
Comment (frits-exmachina) [reply src/LiveryBridge.ts:6]: It was optional because some messages didn't need id's in earlier versions. Now this is no longer necessary.

File: src/LiveryBridge.ts (hunk 8 above)
Comment (frits-exmachina) [reply src/LiveryBridge.ts:118]: Is returning false, when `type` is not a string enough? In that case a reject message is sent back.

Comment (frits-exmachina) [reply src/LiveryBridge.ts:118]: Other type guards already check if `type` is a string. For example: `object.type === 'reject'`

File: src/LiveryBridge.ts (hunk 6 above)
Comment (frits-exmachina) [reply src/LiveryBridge.ts:68]: In that case I'll also pass a boolean `messageIn`, so the messageListener can differentiate between an incoming and outgoing message. Passing a `title` string to be put before the message might also be an idea. What do you think?

File: src/LiveryBridge.ts [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+      this.handleResolveMessage(message);
+    } else if (LiveryBridge.isRejectMessage(message)) {
+      this.handleRejectMessage(message);
+    } else if (LiveryBridge.isCommandMessage(message)) {
+      this.handleCommandMessage(message);
+    } else if (LiveryBridge.isEventMessage(message)) {
+      this.handleEventMessage(message);
+    } else {
+      // should always be last
+      this.sendReject(message.id || 'unknown', 'Unknown Message');
```
Comment (frits-exmachina) [inline src/LiveryBridge.ts:236]: Should we send a reject here? Or just throw an error?

Comment (difosfor) [reply src/LiveryBridge.ts:236]: Just throw an error; do try to make that error specific, e.g: `Unsupported message type: ${type}`
